### PR TITLE
Remove github.com/docker/distribution/registry/client package

### PR DESCRIPTION
**Pull request by Luap99**
Using the github.com/docker/distribution/registry/client package will
import many huge prometheus dependencies, e.g.
 * github.com/prometheus/client_golang/prometheus/promhttp
 * github.com/prometheus/client_golang/prometheus
 * github.com/prometheus/procfs
and even more...

All of these dependencies are completely unused AFAICT but will still end
up in a binary because they are imported transitive.

github.com/docker/distribution/registry/client is only used to check
http errors so I think it makes sense to copy only the required code
into the docker package.

I vendored this commit into podman and it saves over 700KB in binary
size.

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM
@mtrmac @vrothberg PTAL

File: docker/errors.go
Comment (TomSweeneyRedHat) [inline docker/errors.go:67]: @vrothberg and @mtrmac 's call, but this almost seems like this new code should be in a separate file.  I"m going back and forth on it.

Comment (vrothberg) [reply docker/errors.go:67]: I concur. It should be a separate file.

In a second commit, we should also unexport the API and symbols.

**Review by vrothberg — COMMENTED** [review]
Dropping 700k from the binary is sufficiently attractive to me.

We definitely need to add the tests as well but let's wait for @mtrmac's blessing before.

File: docker/errors.go
Comment (mtrmac) [reply docker/errors.go:67]: +1 to both, if this would end up happening.

File: docker/errors_test.go
Comment (mtrmac) [inline docker/errors_test.go:8]: If we are going to copy the code, we should also copy `registry/client/errors_test.go`, or any other relevant tests from the original repo.

**Review by mtrmac — COMMENTED** [review]
As is, I’m not at all sure that this is worth it.
- (None of the `client.*Error` values are explicitly documented API, but they are soft of an API in practice. Though that’s much more the case for `github.com/docker/distribution/registry/api/errcode`, with various uses in c/common, Buildah, Podman.)
- Right now this would remove 700 kB, but soon https://github.com/sigstore/fulcio/blob/main/pkg/api/metrics.go would just drag it all back in, and we would we back where we started.

Basically I think we need to decide whether we are going to make no use of Prometheus (or only an opt-in use) a universal rule, and insist/contribute in upstreams to make that viable. Is that reasonable to aim for?

Alternatively, an (admittedly somewhat desperate) approach might be for the top-level users to use `go.mod`’s `replace` directive to just replace all of the Prometheus client with something dummy that does nothing and has no dependencies. (Compare https://github.com/prometheus/client_golang/issues/196 .) That would allow us to drop that dependency chain, whether upstreams cooperate or not.

---

(Now, moving _just_ the parser could be useful just because it would allow us to integrate better — `registryHTTPResponseToError` is just plumbing to interface with docker/distribution, and with the ability to modify `HandleErrorResponse`, most of that plumbing could almost entirely disappear. But that would be a different PR, with much more focus on interoperability/test coverage.)

**Comment by Luap99** [comment]
My goal is to reduce the podman binary size. As it looks there is really not much to do there, the prometheus code is large and we do not need it. Many dependencies are deeply wired and cannot be removed, this one seems like an easy way to safe space. Most of my other findings will only reduce size by < 100 KB. So I would prefer if we could get this in.

> Right now this would remove 700 kB, but soon https://github.com/sigstore/fulcio/blob/main/pkg/api/metrics.go would just drag it all back in, and we would we back where we started.

Yeah that would be unfortunate.

> (None of the client.*Error values are explicitly documented API, but they are soft of an API in practice. Though that’s much more the case for github.com/docker/distribution/registry/api/errcode, with various uses in c/common, Buildah, Podman.)

Do you fear that this change is breaking consumers?

> Alternatively, an (admittedly somewhat desperate) approach might be for the top-level users to use go.mod’s replace directive to just replace all of the Prometheus client with something dummy that does nothing and has no dependencies. (Compare https://github.com/prometheus/client_golang/issues/196 .) That would allow us to drop that dependency chain, whether upstreams cooperate or not.

I doubt that maintenance would be worth it. It would be just a workaround for the actual issue and it would still take some space for the stub interface.

I leave this up for the maintainers to decide if you want to go forward with this before I apply the requested changes.

**Comment by vrothberg** [comment]
> Right now this would remove 700 kB, but soon https://github.com/sigstore/fulcio/blob/main/pkg/api/metrics.go would just drag it all back in, and we would we back where we started.

Great catch, @mtrmac!  I don't think that it should block this PR here though.  Taming the dependencies will never be finished, I guess, and we can tackle them as they come in.  In case of fulcio, I am sure we can introduce a build-tag upstream to debloat - other callers may benefit as well.

**Comment by mtrmac** [comment]
> I don't think that it should block this PR here though. Taming the dependencies will never be finished, I guess, and we can tackle them as they come in.

There’s a thin line between yak shaving an endless, but clearly useful work, and doing something that is ultimately known not to make a difference. I’m honestly not sure on which side we are here.

So I’m somewhat skeptical but not strongly opposed to this PR, and I’m fine with doing this if there is otherwise consensus that there is worth trying.

>  In case of fulcio, I am sure we can introduce a build-tag upstream to debloat - other callers may benefit as well.

Should we try to do that with docker/distribution = distribution/distribution as well, then? (Maybe even the same build tag?)

**Comment by rhatdan** [comment]
I think we should go forward with this PR, and open an issue with Docker.

**Comment by vrothberg** [comment]
> I think we should go forward with this PR, and open an issue with Docker.

That seems contradictory to me.

>  Should we try to do that with docker/distribution = distribution/distribution as well, then? (Maybe even the same build tag?)

Sounds like a good idea to me.  @Luap99, do you want to give it a shot?

**Comment by Luap99** [comment]
> > Should we try to do that with docker/distribution = distribution/distribution as well, then? (Maybe even the same build tag?)
> 
> Sounds like a good idea to me. @Luap99, do you want to give it a shot?

I am not against trying this but size gained will be less, the current import path seems to be `github.com/docker/distribution/registry/client` -> `github.com/docker/distribution/registry/storage/cache` -> `github.com/docker/distribution/metrics` -> `github.com/docker/go-metrics` -> `github.com/prometheus/client_golang/prometheus` -> and more prometheus stuff

When we add a build tag to disable prometheus, we would only gain this one dependency. In theory the other packages in this list could import other big dependencies that we do not need. This is not be the case today but it could be in the future. 

If we want to change upstream I think it would make more sense to split the error handling into a separate package which does not need further imports and we could import this package instead.

**Comment by tyler92** [comment]
Is there any chance to merge this PR? I find it very useful too.

**Comment by Luap99** [comment]
> Right now this would remove 700 kB, but soon https://github.com/sigstore/fulcio/blob/main/pkg/api/metrics.go would just drag it all back in, and we would we back where we started.

@mtrmac Is the sigstore work fully merged?

I do not see any new references to `github.com/prometheus/client_golang/prometheus` so I think it would make sense to try to remove it from here.

I can also try to push https://github.com/distribution/distribution/pull/3608 forward if this option would be preferred?

**Comment by mtrmac** [comment]
> > Right now this would remove 700 kB, but soon https://github.com/sigstore/fulcio/blob/main/pkg/api/metrics.go would just drag it all back in, and we would we back where we started.
>
> @mtrmac Is the sigstore work fully merged?

The Fulcio part does not exist in c/image yet — but I have also removed that dependency in https://github.com/sigstore/fulcio/pull/616 . So this specific concern no longer applies, and that turns this PR from ineffective to useful. (OTOH some more users of docker/distribution have been added. AFAICS those don’t include the registry/client subpackage right now.)

That makes this PR quite a bit more attractive to me. I read @vrothberg as being supportive of the idea from the start, so, let’s do it.

Could you address the earlier review feedback, please?

---

> I can also try to push https://github.com/distribution/distribution/pull/3608 forward if this option would be preferred?

It would be nice for the wider ecosystem, to spread the idea that simple clients should not have hard-coded dependencies on infrastructure for serving metrics over HTTP. Personally, I wouldn’t say that I _prefer_ that PR any more — d/distribution releases have been very infrequent, and moving the HTTP parser in-package so that we don’t need to override its outcomes in docker/errors.go (as in #1299 ) could — eventually, _after_ we have much more test coverage and confidence — result in a simpler implementation overall.

**Comment by Luap99** [comment]
> Could you address the earlier review feedback, please?

done

**Comment by rhatdan** [comment]
LGTM

File: docker/distribution_error.go
Comment (mtrmac) [inline docker/distribution_error.go:27]:
```suggestion
	dockerChallenge "github.com/docker/distribution/registry/client/auth/challenge"
```
(or just leave it as `challenge` without renaming?)

**Review by mtrmac — COMMENTED** [review]
LGTM other than that one nit, feel free to merge without another review round.

Thanks!

(FWIW, this removes about 3 % from Skopeo’s size on macOS.)

File: docker/distribution_error.go
Comment (Luap99) [reply docker/distribution_error.go:27]: oh yeah lets not rename imports if it is not required

Comment (Luap99) [reply docker/distribution_error.go:27]: looks like we have to rename since challenge is already declared here: https://github.com/containers/image/blob/3aed337418462460c111a3beade6723832af5be0/docker/wwwauthenticate.go#L13
but will fix the typo.

**Review by mtrmac — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**
